### PR TITLE
build and push docker on release

### PR DESCRIPTION
**Pull request by mpnowacki-reef**

### CODE REVIEW

File: test/integration/conftest.py [hunk 1]
```diff
@@ -155,6 +168,21 @@ def b2_tool(global_b2_tool):
     return global_b2_tool
 
 
+@pytest.fixture(autouse=True, scope='session')
+def copy_readme_for_tests():
+    """Copy the README.md file to /tmp so that docker tests can access it"""
+    tmp_readme = pathlib.Path(f'{TEMPDIR}/README.md')
+    if not tmp_readme.exists():
```
Comment (mjurbanski-reef) [inline test/integration/conftest.py:175]: I wanted to comment how this won't always update the readme... but I see now this doesn't matter at all as we do not care about  readme contents as at all as we really ne ANY sample file;

which brings again this would better be served by `sample_file` fixture that would return sample file path - if that file happens to be a readme copied from somewhere wouldn't matter then.

File: test/integration/test_b2_command_line.py [hunk 2]
```diff
@@ -2646,18 +2658,24 @@ def test_cut(b2_tool, bucket_name):
...
 
 
 @skip_on_windows
-def test_upload_unbound_stream__redirect_operator(bash_runner, b2_tool, bucket_name, request):
+def test_upload_unbound_stream__redirect_operator(
+    request, bash_runner, b2_tool, bucket_name, is_running_on_docker
+):
     """Test upload-unbound-stream from stdin using redirect operator."""
+    if is_running_on_docker:
+        pytest.skip('Not supported on Docker')
```
Comment (mjurbanski-reef) [inline test/integration/test_b2_command_line.py:2678]: why not?

File: test/unit/test_console_tool.py [hunk 3]
```diff
@@ -2737,3 +2738,7 @@ def test_rm_skipping_over_errors(self):
         b/b1/test.csv
         '''
         self._run_command(['ls', '--recursive', 'my-bucket'], expected_stdout)
+
+    def test_version(self):
```
Comment (mjurbanski-reef) [inline test/unit/test_console_tool.py:2742]: :+1:

File: test/integration/test_b2_command_line.py [hunk 4]
```diff
@@ -2404,8 +2413,11 @@ def test_replication_monitoring(b2_tool, bucket_name, b2_api):
     )
 
     # make test data
+    shutil.copyfile(
+        pathlib.Path(__file__).parent.parent.parent / 'CHANGELOG.md', f'{TEMPDIR}/CHANGELOG.md'
+    )
```
Comment (mjurbanski-reef) [inline test/integration/test_b2_command_line.py:2418]: seems like we have two different places in which we create these sample files

File: test/integration/helpers.py [hunk 5]
```diff
@@ -454,6 +407,78 @@ def should_succeed(
...
+    def execute(
+        self,
+        args: Optional[List[Union[str, Path, int]]] = None,
+        additional_env: Optional[dict] = None,
+    ):
+        """
+        :param cmd: a command to run
+        :param args: command's arguments
+        :param additional_env: environment variables to pass to the command, overwriting parent process ones
+        :param env_file_cmd_placeholder: If specified, all occurrences of this string in `args` will be substituted with a
```
Comment (mjurbanski-reef) [inline test/integration/helpers.py:441]: no such argument

File: test/integration/test_b2_command_line.py [hunk 6]
```diff
@@ -1189,9 +1196,14 @@ def test_sse_b2(b2_tool, bucket_name):
...
 
 
-def test_sse_c(b2_tool, bucket_name):
+def test_sse_c(b2_tool, bucket_name, is_running_on_docker):
 
-    file_to_upload = 'README.md'
+    sse_c_key_id = 'user-generated-key-id \nąóźćż\nœøΩ≈ç\nßäöü'
+    if is_running_on_docker:
+        # TODO: fix this once we figure out how to pass env vars with \n in them to docker, docker-compose should work
+        sse_c_key_id = sse_c_key_id.replace('\n', '')
```
Comment (mjurbanski-reef) [inline test/integration/test_b2_command_line.py:1204]:
```
export TEST_VAR=$(echo "a\nb")
docker run --rm -e TEST_VAR ubuntu env
```
this will require dropping env-file support and replacing it with `-e ENVVAR` listing, but what can you do.
If you don't like idea of putting docker-specific solutions into pytest (which we already do with skips), we can potentially write a `docker_env_run.sh` that would prepare -e `$(env | awk -F "=" '{printf "-e %s=%s ", $1, $2}'` inside of it and pytest would call it thru `--sut` as any other command

File: test/integration/test_b2_command_line.py [hunk 7]
```diff
@@ -212,38 +216,39 @@ def test_basic(b2_tool, bucket_name):
...
-    log_file_regex = re.compile(
-        r'urllib3.connectionpool\tDEBUG\t.* "POST /b2api/v2/b2_delete_bucket HTTP'
-        r'.*' + stack_trace_in_log,
-        re.DOTALL,
-    )
-    with open('b2_cli.log') as logfile:
-        log = logfile.read()
-        assert re.search(log_file_regex, log), log
-    os.remove('b2_cli.log')
+    if not is_running_on_docker:  # It's difficult to read the log in docker in CI
```
Comment (mjurbanski-reef) [inline test/integration/test_b2_command_line.py:219]: looking at this and README problem, I wonder if mounting local folder as workdir in docker would kill both birds with one stone
using `-u $(id -u):$(id -g)` it make it also pretty safe to do as well

File: README.md [hunk 8]
```diff
@@ -117,26 +127,33 @@ Note that using many threads could in some cases be detrimental to the other use
 
 ### Docker image
 
-An official Docker image is provided for these who want to use B2 Command Line Tool in a Docker environment.
+#### Authorization
+
+User can either authorize on each call (`list-buckets` is just an exemplary command here)
```
Comment (mjurbanski-reef) [inline README.md:132]:
```suggestion
User can either authorize on each command (`list-buckets` is just a example here)
```

**Review by mjurbanski-reef — COMMENTED** [review]
nice work

File: test/integration/test_b2_command_line.py (hunk 2 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:2678]: To the best of my knowledge this can’t work. Running it yields strange errors about file descriptors not existing on the container os

File: test/integration/test_b2_command_line.py [hunk 9]
```diff
@@ -61,7 +64,7 @@ def get_bucketinfo() -> Tuple[str, str]:
 
 def test_download(b2_tool, bucket_name):
 
-    file_to_upload = 'README.md'
+    file_to_upload = f'{TEMPDIR}/README.md'
```
[the remark this answers is not in the document]
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:67]: good idea

File: test/integration/conftest.py (hunk 1 above)
Comment (mpnowacki-reef) [reply test/integration/conftest.py:175]: agreed

File: test/integration/test_b2_command_line.py (hunk 4 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:2418]: good catch!

File: test/integration/test_b2_command_line.py (hunk 6 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:1202]: honestly I'm not sure this is worth, do you think so? It's a rare corner case, that probably doesn't require testing anyway

File: test/integration/test_b2_command_line.py (hunk 7 above)
Comment (mpnowacki-reef) [reply test/integration/test_b2_command_line.py:212]: I have considered this and don't like this approach, because it makes the tests work in significantly different environment than the target one. Users will not necessarily want to mount container user's homedir to a host directory. Before this PR, there were docker tests, and they worked by copying test code to the container and running everything in the container. Tests passed, but just running `docker run b2 ls bucket` failed, as the container user didn't have read rights to his own homedir, which flew under the radar of all tests. The tested image and the released image were slightly different, enough to invalidate all integration tests.

File: test/integration/test_b2_command_line.py (hunk 2 above)
Comment (mjurbanski-reef) [reply test/integration/test_b2_command_line.py:2674]: oh, right. thanks

File: test/integration/test_b2_command_line.py (hunk 6 above)
Comment (mjurbanski-reef) [reply test/integration/test_b2_command_line.py:1202]: I'm just scared about someone loosing a ton of time on debugging this when they actually need it.

I guess solution in the middle would be to emit logger warning in https://github.com/reef-technologies/B2_Command_Line_Tool/pull/213/files#diff-0457e1c77e315509ddcb03e30c7d249298c33f012541c7e3654b6728e7cf56b5R424 IF any of the vars contains newlines. That way when adding a new test it blows up the warning will be visible in pytest output.

**Review by mjurbanski-reef — APPROVED** [review]
